proposals/simd-0170-builtin-instruction-cost-and-budget.md

```diff
@@ -0,0 +1,127 @@
+being executed when they should fail due to exceeding the budget. Consequently,
+the cost tracker may need to account for additional compute units toward block
+limits after transaction execution, potentially producing blocks that exceed
+those limits.
+
+Furthermore, maintaining consistency in transaction costs between the banking
+stage and SVM would simplify the code logic and make reasoning more
+straightforward.
+
+## Alternatives Considered
+
+- One possible alternative approach would be to maintain the current allocation
+of compute budget for builtin instructions but add logic to the cost tracker
+to account for the discrepancy during tracking. However, this would add
+complexity and could introduce additional corner cases, potentially leading to
+more issues.
+
+- Another alternative would be to treat builtin instructions the same as other
+instructions by allocating DEFAULT_INSTRUCTION_COMPUTE_UNIT_LIMIT units for them
+as well. However, this approach has concerns. DEFAULT_INSTRUCTION_COMPUTE_UNIT_LIMIT
+is a very conservative estimate, currently set at 200,000 CUs per instruction.
+Estimating all builtin instructions, including votes and transfers, would cause
+the banking stage to significantly over-reserve block space during block
+production, potentially leading to under-packed blocks. Additionally, if it's
+known that builtin instructions will consume a fixed amount of CUs, it doesn't
+make sense to estimate them with a generic DEFAULT_INSTRUCTION_COMPUTE_UNIT_LIMIT.
+
+## New Terminology
+
+None
+
+## Detailed Design
+
+To ensure consistency, the following three changes are proposed:
+
+1. Allocate Static Compute Budget for Builtin Instructions:
+   When the compute-unit-limit is not explicitly requested, the compute budget
+should always allocate the statically defined DEFAULT_COMPUTE_UNITS for builtin
+instructions, including compute-budget instructions.
+
+   Currently, when set_compute_unit_limit is not used, all instructions are
+allocated DEFAULT_INSTRUCTION_COMPUTE_UNIT_LIMIT, capped by the transaction's
+MAX_COMPUTE_UNIT_LIMIT. However, compute-budget instructions historically
+haven't had any units allocated. This can lead to over-packed blocks in
+certain scenarios.
+
+   [Example 1](https://github.com/anza-xyz/agave/pull/2746/files#diff-c6c8658338536afbf59d65e9f66b71460e7403119ca76e51dc9125e1719f4f52R13403-R13429):
+   A transaction consists of a Transfer and an expensive non-builtin instruction.
+If the non-builtin instruction requires more than DEFAULT_INSTRUCTION_COMPUTE_UNIT_LIMIT
+CUs, the over-budgeting for Transfer allows the expensive instruction to
+execute to completion, forcing an upward adjustment.
+
+   [Example 2](https://github.com/anza-xyz/agave/pull/2746/files#diff-c6c8658338536afbf59d65e9f66b71460e7403119ca76e51dc9125e1719f4f52R13431-R13455):
+   A builtin instruction that might call other instructions (CPI) would fail
+without explicitly requesting more CUs. However, this isn't currently happening.
```

Comment:
Under this proposal, this transaction would fail, right?

That seems to indicate to me that cost-model is broken for builtins (even with this proposal).
If we are claiming these builtins have known static costs, we should use those costs instead of creating a weird UX where some native/builtin instructions **require** the use of some other one.

In addition to this, why can we not fix the cost-tracking done in the VM?
If we are going to continue claiming in cost-model that some builtin program always uses X CUs, then that's what it should use - regardless of internal behavior which may CPI.

Reply:
> Under this proposal, this transaction would fail, right?

Yes, it'd fail under this proposal, kinda of intentionally.

> If we are claiming these builtins have known static costs, we should use those costs instead of creating a weird UX where some native/builtin instructions require the use of some other one.

Yea, builtins that CPI to other instructions is a tricky case, perhaps it can't be said to have a static cost, unless the number of CPIs are static. I don't have a good immediate solution for it, but asking/pushing for every transaction to explicitly call set_compute_unit_limit.

> In addition to this, why can we not fix the cost-tracking done in the VM?

To reduce confusion, let's use "cost tracking" for block packing side of business, and "budget consuming" when relates to VM takes CUs.

Reply:
> In addition to this, why can we not fix the cost-tracking done in the VM?
> If we are going to continue claiming in cost-model that some builtin program always uses X CUs, then that's what it should use - regardless of internal behavior which may CPI.

Technically the VM is doing its designated job correctly, since the `InvokeContext` is created per-instruction, and it handles the accrual of CU usage across CPI stack frames.

I feel like the solution would have to be at the SVM (runtime) level, where we keep some list of builtin CUs that are reserved for each instruction. Something like this might do it (sorry, bunch of pseudo-code here).

```rust
// ALT program:
mod address_lookup_table_program {
    const CUS_BASE_CREATE_LOOKUP_TABLE: u64 = 750;
    const CUS_TOTAL_CREATE_LOOKUP_TABLE: u64 = 750 + 150 + 150; // Includes CPI to System.

    enum Instruction {
        CreateLookupTable,
        /* ... */
    }

    impl Instruction {
        fn get_cus_to_allocate(ix_data: &[u8]) -> BuiltinComputeUnits {
            match ix_data.first() {
                Some(&0) => CUS_TOTAL_CREATE_LOOKUP_TABLE,
                /* ... */
            }
        }
    }

    // Processor for `CreateLookupTable`:
    fn process_create_lookup_table(/* ... */) {
        // Base CUs. CPIs will consume their corresponding base CUs.
        consume_cus(CUS_BASE_CREATE_LOOKUP_TABLE);
    }
}


// SVM:
fn process_tx(tx: SVMTransaction, /* ... */) {
    let cus_to_allocate = tx
        .instructions()
        .iter()
        .map(|ix| {
            if ix.program_id == address_lookup_table_program::id() {
                address_lookup_table_program::Instruction::get_cus_to_allocate(ix.data)
            }
            // Default allocation 200k.
            200_000
        })
        .sum();
    allocate_cus_for_transaction(cus_to_allocate);
}
```

Reply:
There's possibility "builtin instruction" can be deprecated all together. But if it stays, then allocating budget by builtin "instructions" not by "program" would be next step. ZK program [consumes CUs by instruction](https://github.com/anza-xyz/agave/blob/master/programs/zk-token-proof/src/lib.rs#L186), not by program_id.

Reply:
> ZK program [consumes CUs by instruction](https://github.com/anza-xyz/agave/blob/master/programs/zk-token-proof/src/lib.rs#L186), not by program_id.

Awesome! I'm in favor.